src/datadog/tracer_telemetry.cpp

```diff
@@ -0,0 +1,211 @@
+#include "tracer_telemetry.h"
+
+#include "logger.h"
+#include "platform_util.h"
+#include "span_defaults.h"
+#include "version.h"
+
+namespace datadog {
+namespace tracing {
+
+TracerTelemetry::TracerTelemetry(
+    bool enabled, const Clock& clock, const std::shared_ptr<Logger>& logger,
+    const std::shared_ptr<const SpanDefaults>& span_defaults,
+    const RuntimeID& runtime_id)
+    : enabled_(enabled),
+      clock_(clock),
+      logger_(logger),
+      span_defaults_(span_defaults),
+      runtime_id_(runtime_id),
+      hostname_(get_hostname().value_or("hostname-unavailable")) {
+  if (enabled_) {
```

Comment:
It is probably better to move that logic in the caller, so depending of `enabled` it can instantiate a `NoopTracerTelemetry` or this implementation.

Reply:
I'm thinking about what a `TracerTelemetry` _interface_ would look like. What would `metrics()` return?

I prefer to leave this concrete as Caleb has it, but Damien's suggestion is worth thinking about more.